engine/executor/dm/task.go

```diff
@@ -0,0 +1,253 @@
+	onFinished(ctx context.Context) error
+	createUnitHolder(cfg *config.SubTaskConfig) unit.Holder
+}
+
+// baseTask implements some default methods for dm task
+type baseTask struct {
+	task
+	lib.BaseWorker
+	unitHolder   unit.Holder
+	messageAgent dmpkg.MessageAgent
+
+	ctx                context.Context
+	cancel             context.CancelFunc
+	mu                 sync.RWMutex
+	cfg                *dmconfig.SubTaskConfig
+	storageWriteHandle broker.Handle
+	stage              metadata.TaskStage
+	workerType         libModel.WorkerType
+	taskID             string
+	masterID           libModel.MasterID
+}
+
+// newBaseTask creates BaseDMTask instances
+func newBaseTask(dCtx *dcontext.Context, masterID libModel.MasterID, workerType libModel.WorkerType, conf lib.WorkerConfig) *baseTask {
+	ctx, cancel := context.WithCancel(context.Background())
+	return &baseTask{
+		ctx:        ctx,
+		cancel:     cancel,
+		cfg:        conf.(*config.SubTaskConfig),
+		stage:      metadata.StageInit,
+		workerType: workerType,
+		taskID:     conf.(*config.SubTaskConfig).SourceID,
+		masterID:   masterID,
+	}
+}
+
+func (t *baseTask) createComponents(ctx context.Context) error {
+	log.L().Debug("create components")
+	t.messageAgent = dmpkg.NewMessageAgent(t.ctx, map[string]dmpkg.Sender{t.masterID: t}, t)
+	return nil
+}
+
+// InitImpl implements lib.WorkerImpl.InitImpl
+func (t *baseTask) InitImpl(ctx context.Context) error {
+	log.L().Info("init task")
+	if err := t.createComponents(ctx); err != nil {
+		return err
+	}
+	if err := t.onInit(ctx); err != nil {
+		return err
+	}
+	t.unitHolder = t.createUnitHolder(t.cfg)
+	return t.unitHolder.Init(ctx)
+}
+
+// Tick implements lib.WorkerImpl.Tick
+func (t *baseTask) Tick(ctx context.Context) error {
+	t.unitHolder.LazyProcess(ctx)
+	if err := t.unitHolder.Tick(ctx); err != nil {
+		return err
+	}
+	return t.tryUpdateStatus(ctx)
+}
+
+// Workload implements lib.WorkerImpl.Worload
+func (t *baseTask) Workload() model.RescUnit {
+	log.L().Info("dmtask.Workload")
+	return 0
+}
+
+// OnMasterFailover implements lib.WorkerImpl.OnMasterFailover
+func (t *baseTask) OnMasterFailover(reason lib.MasterFailoverReason) error {
+	log.L().Info("dmtask.OnMasterFailover")
+	return nil
+}
+
+// OnMasterMessage implements lib.WorkerImpl.OnMasterMessage
+func (t *baseTask) OnMasterMessage(topic p2p.Topic, message p2p.MessageValue) error {
+	log.L().Info("dmtask.OnMasterMessage", zap.String("topic", topic), zap.Any("message", message))
+	return t.messageAgent.OnMessage(t.masterID, topic, message)
+}
+
+// CloseImpl implements lib.WorkerImpl.CloseImpl
+func (t *baseTask) CloseImpl(ctx context.Context) error {
+	t.cancel()
+	t.unitHolder.Close()
+	return nil
+}
+
+// exit closes the task and exits.
+func (t *baseTask) exit(ctx context.Context, status libModel.WorkerStatus, err error) error {
+	if err := t.CloseImpl(ctx); err != nil {
+		log.L().Warn("fail to close task", log.ShortError(err))
+	}
+	return t.Exit(ctx, status, err)
```

Comment:
it's weird that a private "exit" calls public "Exit". To solve this problem, maybe rename it closeAndExit or the framework should call CloseImpl in Exit? @liuzix